File: week05/BOJ_18119_단어암기/BOJ_18119_dooyeong.py
from sys import stdin

N, M = map(int, stdin.readline().split())
wordList = [0] * N

for i in range(N):
    word = stdin.readline().strip()
    tmp = 0
    a = ord('a')

    for w in word:
        tmp |= 1 << ord(w) - a

    wordList[i] = tmp

status = (1 << 27) - 1
aeiou = set('aeiou')

for _ in range(M):
    cmd, ch = stdin.readline().split()

    if ch in aeiou:
        continue

    if cmd == '1':
        status &= ~(1 << (ord(ch) - ord('a')))
    else:
        status |= (1 << (ord(ch) - ord('a')))

    cnt = 0

    for i in range(N):
        if wordList[i] & status == wordList[i]:
            cnt += 1

    print(cnt)

File: week05/BOJ_18119_단어암기/BOJ_18119_dooyeong.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().split("\n");
const [N, M] = input
  .shift()
  .split(" ")
  .map((x) => +x);

const words = [],
  a = "a".charCodeAt(0);

const getWordBit = (word) => {
  let tmp = 0;

  for (let i = 0, l = word.length; i < l; i++) {
    tmp |= 1 << (word[i].charCodeAt(0) - a);
  }

  return tmp;
};

const getCharBit = (ch) => {
  return 1 << (ch.charCodeAt(0) - a);
};

for (let i = 0; i < N; i++) {
  words[i] = getWordBit(input[i]);
}

let status = (1 << 27) - 1,
  ans = "";

for (let i = N; i < N + M; i++) {
  const [cmd, ch] = input[i].split(" ");

  if (cmd == "1") status &= ~getCharBit(ch);
  else status |= getCharBit(ch);

  let cnt = 0;
  for (let i = 0; i < N; i++) {
    if ((words[i] & status) == words[i]) cnt += 1;
  }

  ans += cnt.toString() + "\n";
}

console.log(ans);
